core/src/test/java/org/apache/calcite/test/LargeGeneratedJoinTest.java

```diff
@@ -0,0 +1,156 @@
+import org.apache.calcite.schema.Schema;
+import org.apache.calcite.schema.SchemaPlus;
+import org.apache.calcite.schema.Table;
+import org.apache.calcite.schema.impl.AbstractSchema;
+import org.apache.calcite.sql.parser.SqlParseException;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.tools.RelConversionException;
+import org.apache.calcite.tools.ValidationException;
+
+import com.google.common.collect.ImmutableMap;
+
+import org.junit.jupiter.api.Test;
+
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/**
+ * Test class for CALCITE-3094.
+ */
+public class LargeGeneratedJoinTest {
+
+  /**
+   * Marker interface for Field.
+   */
+  interface FieldT extends BiConsumer<RelDataTypeFactory, RelDataTypeFactory.Builder> {
+  }
+
+  /**
+   * Marker interface for Row.
+   */
+  interface RowT extends Function<RelDataTypeFactory, RelDataType> {
+  }
+
+  static FieldT field(String name) {
+    return (tf, b) -> b.add(name, SqlTypeName.VARCHAR);
+  }
+
+  static RowT row(FieldT... fields) {
+    return tf -> {
+      RelDataTypeFactory.Builder builder = tf.builder();
+      for (FieldT f : fields) {
+        f.accept(tf, builder);
+      }
+      return builder.build();
+    };
+  }
+
+  private static QueryableTable tab(int fieldCount) {
+    List<Row> lRow = new ArrayList<>();
+    for (int r = 0; r < 2; r++) {
+      Object[] current = new Object[fieldCount];
+      for (int i = 0; i < fieldCount; i++) {
+        current[i] = "v" + i;
+      }
+      lRow.add(Row.of(current));
+    }
+
+    List<FieldT> fields = new ArrayList<>();
+    for (int i = 0; i < fieldCount; i++) {
+      fields.add(field("F_" + i));
+    }
+
+    final Enumerable<?> enumerable = Linq4j.asEnumerable(lRow);
+    return new AbstractQueryableTable(Row.class) {
+
+      @Override public RelDataType getRowType(RelDataTypeFactory typeFactory) {
+        return row(fields.toArray(new FieldT[fieldCount])).apply(typeFactory);
+      }
+
+      @Override public <T> Queryable<T> asQueryable(QueryProvider queryProvider, SchemaPlus schema,
+          String tableName) {
+        return (Queryable<T>) enumerable.asQueryable();
+      }
+    };
+  }
+
+  @Test public void test() throws SqlParseException, RelConversionException, ValidationException {
+    Schema rootSchema = new AbstractSchema() {
+      @Override protected Map<String, Table> getTableMap() {
+        return ImmutableMap.of("T0", tab(100),
+            "T1", tab(101));
+      }
+    };
+
+    final CalciteSchema sp = CalciteSchema.createRootSchema(false, true);
+    sp.add("ROOT", rootSchema);
+
+    String sql = "SELECT * \n"
+        + "FROM ROOT.T0 \n"
+        + "JOIN ROOT.T1 \n"
+        // + "ON ROOT.T0.F_0 = ROOT.T1.F_0";
```

Comment:
if this is not used maybe it can be deleted

Reply:
Removed.